pkg/auth/oauth/registry/timeoutvalidator.go

```diff
@@ -0,0 +1,267 @@
+		// FIXME: what is the right degree for the btree
+		btree.New(32),
+		timeoutAsDuration(defaultTimeout),
+		timeoutAsDuration(flushTimeout),
+		timeoutAsDuration(safetyMargin),
+	}
+	glog.V(5).Infof("Token Timeout Updater primed with defaultTimeout=%d flushTimeout=%d safetyMargin=%d", defaultTimeout, flushTimeout, safetyMargin)
+	return ttu
+}
+
+func timeoutAsDuration(timeout int32) time.Duration {
+	return time.Duration(timeout) * time.Second
+}
+
+// CheckTimeout is called with a token when it is seen by an authenticator
+// it touches only the tokenChannel so it is safe to call from other threads
+func (a *TimeoutValidator) Validate(token *oauth.OAuthAccessToken, _ *user.User) error {
+	if token.TimeoutsIn == 0 {
+		// We care only if the token was created with a timeout to start with
+		return nil
+	}
+
+	td := tokenData{
+		token: token,
+		seen:  time.Now(),
+	}
+	if td.timeout().Before(td.seen) {
+		return ErrTimedout
+	}
+
+	// After a positive timeout check we need to update the timeout and
+	// schedule an update so that we can either set or update the Timeout
+	// we do that launching a micro goroutine to avoid blocking
+	go a.updateToken(td)
+
+	return nil
+}
+
+func (a *TimeoutValidator) updateToken(td tokenData) {
+	a.tokenChannel <- td
+}
+
+func (a *TimeoutValidator) updateTimeouts(oauthClient *oauth.OAuthClient) {
+	timeout := *oauthClient.AccessTokenTimeoutSeconds / 3
+	flushTimeout := int32(a.flushTimeout / time.Second)
+	if timeout < flushTimeout {
+		if timeout < validation.MinFlushTimeout {
+			timeout = validation.MinFlushTimeout
+		}
+		a.flushTimeout = timeoutAsDuration(timeout)
+		a.safetyMargin = timeoutAsDuration(timeout / 10)
+		glog.V(5).Infof("OauthClient %q updates timeouts to flushTimeout=%s safetyMargin=%s", oauthClient.Name, a.flushTimeout, a.safetyMargin)
+	}
+}
+
+func (a *TimeoutValidator) clientTimeout(name string) time.Duration {
+	oauthClient, err := a.oauthClient.Get(name)
+	if err != nil {
+		glog.V(5).Infof("Failed to fetch OAuthClient %q for timeout value: %v", name, err)
+		return a.defaultTimeout
+	}
+	if oauthClient.AccessTokenTimeoutSeconds == nil {
+		return a.defaultTimeout
+	}
+	a.updateTimeouts(oauthClient)
+	return timeoutAsDuration(*oauthClient.AccessTokenTimeoutSeconds)
+}
+
+func (a *TimeoutValidator) insert(td tokenData) {
+	// if there is a token we have to remove it first
+	oldtd, exists := a.data[td.token.Name]
+	if exists {
+		a.delete(oldtd, &tokenDataRef{oldtd.token.Name, oldtd.timeout()})
+	}
+	a.data[td.token.Name] = td
+	a.tree.ReplaceOrInsert(&tokenDataRef{td.token.Name, td.timeout()})
+}
+
+func (a *TimeoutValidator) delete(td tokenData, tdr *tokenDataRef) {
+	a.tree.Delete(tdr)
+	delete(a.data, td.token.Name)
+}
+
+func (a *TimeoutValidator) update(td tokenData) error {
+	delta := a.clientTimeout(td.token.ClientName)
+	newTimeout := int32(0)
+	if delta > 0 {
+		newTimeout = int32((td.seen.Sub(td.token.CreationTimestamp.Time) + delta) / time.Second)
```

Comment:
I had some comments explaining this math which I think are useful.

Reply:
added